crates/subspace-service/src/sync_from_dsn/snap_sync.rs

```diff
@@ -292,9 +298,46 @@ where
         let reconstructor = Arc::new(Mutex::new(Reconstructor::new(erasure_coding.clone())));
 
         for segment_index in segments_to_reconstruct {
-            let segment_pieces = download_segment_pieces(segment_index, piece_getter)
-                .await
-                .map_err(|error| format!("Failed to download segment pieces: {error}"))?;
+            let mut segment_pieces = None;
+            let mut segment_error = Ok(());
+
+            for retry in 0..SEGMENT_DOWNLOAD_RETRIES {
+                // Download segment pieces
+                let result = download_segment_pieces(segment_index, piece_getter).await;
+
+                match result {
+                    Ok(pieces) => {
+                        segment_pieces = Some(pieces);
+                        segment_error = Ok(());
+                        break;
+                    }
+                    Err(error) => {
+                        if retry == SEGMENT_DOWNLOAD_RETRIES - 1 {
+                            error!(
+                                target: LOG_TARGET,
+                                %error,
+                                %retry,
+                                "Failed to download segment pieces"
+                            );
+                        } else {
+                            warn!(
+                                target: LOG_TARGET,
+                                %error,
+                                %retry,
+                                ?SEGMENT_DOWNLOAD_RETRY_DELAY,
+                                "Failed to download segment pieces, retrying..."
+                            );
+                        }
+                        segment_error = Err(error);
+                        sleep(SEGMENT_DOWNLOAD_RETRY_DELAY).await
+                    }
+                }
+            }
+
+            if let Err(segment_error) = segment_error {
```

Comment:
Both `segment_pieces` and `segment_error` are in sync with each other, so there should probably be just one variable that stores `Result`. `match` here will handle `Err` case and there will be no need to use `.expect()` below at all.